### PR TITLE
fix up shred sigverify stats

### PR DESCRIPTION
**Pull request by bw-solana**
#### Problem
`fn run_shred_sigverify` increments stats for where we are losing packets (e.g. discarded after sigverify). However, it does not properly subtract out discards from previous stages. E.g. `num_discards_post` includes `num_discards_pre` and `num_duplicates` in its count.

The unit test that was added fails on the original code.

#### Summary of Changes

- Subtract out discards from previous checks when tallying up stats
- Add a unit test to make sure counts increment as expected

### CODE REVIEW

File: turbine/src/sigverify_shreds.rs [hunk 1]
```diff
@@ -137,7 +139,8 @@ fn run_shred_sigverify<const K: usize>(
         &mut packets,
         cache,
     );
-    stats.num_discards_post += count_discards(&packets);
+    let num_discards_post = count_discards(&packets) - num_duplicates - num_discards_pre;
```
Comment (bw-solana) [inline turbine/src/sigverify_shreds.rs:142]: This is the key change. Subtracting counts from pre/duplicate

File: turbine/src/sigverify_shreds.rs [hunk 2]
```diff
@@ -377,4 +380,63 @@ mod tests {
         assert!(!batches[0][0].meta().discard());
         assert!(batches[0][1].meta().discard());
     }
+
+    fn run_shred_sigverify_for_test(
```
Comment (bw-solana) [inline turbine/src/sigverify_shreds.rs:384]: Open to a better name for this function

File: turbine/src/sigverify_shreds.rs [hunk 3]
```diff
@@ -377,4 +380,63 @@ mod tests {
...
+        // Send packet batch into the channel.
+        shred_fetch_sender.send(packet_batch).unwrap();
+
+        // Handle verification of this packet.
+        run_shred_sigverify_for_test(&shred_fetch_receiver, &mut stats);
+
+        // Ensure discard stats incremented as expected.
+        assert!(stats.num_discards_pre == 1);
+        assert!(stats.num_duplicates == 0);
+        assert!(stats.num_discards_post == 0);
```
Comment (bw-solana) [inline turbine/src/sigverify_shreds.rs:425]: This is the assert that fails without the changes to `run_shred_sigverify`

**Comment by behzadnouri** [comment]
This is what `pre/post` intended to mean.
* `shred_sigverify.num_discards_pre` was meant to indicate number of packets marked as `discard` _before_ sigverify stage.
* `shred_sigverify.num_discards_post` was meant to indicate number of packets marked as `discard` _after_ sigverify stage.

So number of packets discarded during sigverify is `num_discards_post - num_discards_pre` of which `num_duplicates` were  duplicates.

**Comment by bw-solana** [comment]
> So number of packets discarded during sigverify is `num_discards_post - num_discards_pre` of which `num_duplicates` were duplicates.

@behzadnouri Ah, gotcha, so they're cumulative by intent. What should we do with this PR? Just abandon it, rework the tests to match what you mentioned, something else?

**Comment by behzadnouri** [comment]
> What should we do with this PR? Just abandon it, rework the tests to match what you mentioned, something else?

either way is fine with me. your call.

**Pull request closed without merging**